theano/tensor/signal/downsample.py

```diff
@@ -343,17 +366,39 @@ def make_node(self, x, maxout, gz):
     def perform(self, node, inp, out):
         x, maxout, gz = inp
         gx_stg, = out
-        gx = numpy.zeros_like(x)
-
+        
         #number of pooling output rows
         pr = maxout.shape[-2]
         #number of pooling output cols
         pc = maxout.shape[-1]
         ds0, ds1 = self.ds
         st0, st1 = self.st
-        img_rows = x.shape[-2]
-        img_cols = x.shape[-1]
-
+        img_rows = x.shape[-2] + 2 * self.padding[0]
+        img_cols = x.shape[-1] + 2 * self.padding[1]
+        pad_h = self.padding[0]
+        pad_w = self.padding[1]
+        def pad_img(x):
+            w = x.shape[3]
+            h = x.shape[2]
+            fill = x.min()-1
+            t = numpy.ones((x.shape[0],x.shape[1],1,1))
+            ud_bar = (numpy.zeros((pad_h, w)) + fill)[
+                numpy.newaxis, numpy.newaxis,:,:] * t
+            lr_bar = (numpy.zeros((pad_h * 2 + h, pad_w)) + fill)[
+                numpy.newaxis, numpy.newaxis,:,:] * t
+            y = numpy.concatenate([ud_bar, x, ud_bar], axis=2)
+            y = numpy.concatenate([lr_bar, y, lr_bar], axis=3)
+            return y
+        def unpad(g):
+            w = x.shape[3]
+            h = x.shape[2]
+            r_st = pad_h
+            r_end = h + pad_h
+            c_st = pad_w
+            c_end = w + pad_w
+            return g[:,:,r_st:r_end,c_st:c_end]
```

Comment:
You don't really need two auxiliary functions for that, just index by `[:, :, pad_h:-pad_h, pad_w:-pad_w]`